File: bee_five/lib/xp_service.dart
import 'dart:math' as math;

import 'package:shared_preferences/shared_preferences.dart';

import 'adventure_progress_service.dart' show scheduleProgressCloudSync;

/// Pref keys (must match dashboard_page.dart).
const String _prefUserXp = 'user_xp';
const String _prefLoginStreak = 'login_streak';

/// Default XP for new users.
const int defaultXp = 10;

/// XP for 3 consecutive wins in classic game.
const int xpClassicThreeWins = 2;

/// XP for winning a hard practice game.
const int xpHardPracticeWin = 1;

/// XP for winning an adventure game/match.
const int xpAdventureMatchWin = 1;

/// XP for clearing adventure levels 10, 20, 30, …
const int xpAdventureMilestoneLevelWin = 3;

/// XP lost per adventure game loss (each loss = -1 XP).
const int xpAdventureOneLoss = 1;

/// Frontier-level failures before offering a rewarded ad to skip to the next level.
const int adventureLossesBeforeSkipAdOffer = 4;

/// School lobby multiplayer: win / loss delta (loss uses [removeXp], clamped at 0).
const int xpSchoolLobbyMatchDelta = 1;

/// XP awarded after watching a rewarded ad on the home map / Gain XP flows.
const int xpRewardedAdWatch = 2;

const String _prefLastLoginDate = 'last_login_date';
const String _prefDailyChallengeDate = 'daily_challenge_date';
const String _prefDailyChallengeWon = 'daily_challenge_won';
const String _prefAdventureConsecutiveLosses = 'adventure_consecutive_losses';
const String _prefAdventureConsecutiveWins = 'adventure_consecutive_wins';
const String _prefAdventureHighestUnlockedLevel = 'adventure_highest_unlocked_level';
const String _prefAdventureCurrentLevel = 'adventure_current_level';
const String _prefAdventureLevelsFirstClearXp = 'adventure_levels_first_clear_xp';
const String _prefAdventureFirstClearXpMigrated = 'adventure_first_clear_xp_migrated';

/// Top of adventure progression: max of selected level and highest unlocked (prefs can lag).
Future<int> _effectiveAdventureFrontierLevel() async {
  final prefs = await SharedPreferences.getInstance();
  final current = prefs.getInt(_prefAdventureCurrentLevel) ?? 1;
  final highest = prefs.getInt(_prefAdventureHighestUnlockedLevel) ?? current;
  return math.max(current, highest);
}

Future<bool> _isAdventureFrontierLevel(int levelJustPlayedOrCompleted) async {
  final frontier = await _effectiveAdventureFrontierLevel();
  // Win/loss XP at the "top" level the player is on (current or highest, whichever is greater).
  return levelJustPlayedOrCompleted == frontier;
}

/// One-time: treat all levels strictly below the current top as already rewarded for first-clear
/// (+1) so existing saves do not mass-award XP on upgrade.
Future<void> _ensureAdventureFirstClearXpMigrated() async {
  final prefs = await SharedPreferences.getInstance();
  if (prefs.getBool(_prefAdventureFirstClearXpMigrated) == true) return;

  final current = prefs.getInt(_prefAdventureCurrentLevel) ?? 1;
  final highest = prefs.getInt(_prefAdventureHighestUnlockedLevel) ?? current;
  final top = math.max(current, highest);

  final existing = prefs.getStringList(_prefAdventureLevelsFirstClearXp)?.toSet() ?? <String>{};
  for (int i = 1; i < top; i++) {
    existing.add(i.toString());
  }
  final list = existing.toList()..sort((a, b) => int.parse(a).compareTo(int.parse(b)));
  await prefs.setStringList(_prefAdventureLevelsFirstClearXp, list);
  await prefs.setBool(_prefAdventureFirstClearXpMigrated, true);
  scheduleProgressCloudSync();
}

/// Ensures XP is initialized to [defaultXp] if never set.
Future<void> ensureXpInitialized() async {
  final prefs = await SharedPreferences.getInstance();
  if (prefs.getInt(_prefUserXp) == null) {
    await prefs.setInt(_prefUserXp, defaultXp);
  }
}

/// Call when app is opened (e.g. home page init). Updates login streak only.
Future<void> onAppOpen() async {
  final prefs = await SharedPreferences.getInstance();
  await ensureXpInitialized();

  final now = DateTime.now();
  final today = '${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}';
  final last = prefs.getString(_prefLastLoginDate);

  if (last == today) return;

  int streak = prefs.getInt(_prefLoginStreak) ?? 0;

  if (last == null) {
    streak = 1;
  } else {
    final lastDate = DateTime.tryParse(last);
    if (lastDate != null) {
      final diff = now.difference(lastDate).inDays;
      if (diff == 1) {
        streak += 1;
      } else if (diff > 1) {
        streak = 1;
      }
    } else {
      streak = 1;
    }
  }

  await prefs.setInt(_prefLoginStreak, streak);
  await prefs.setString(_prefLastLoginDate, today);
  scheduleProgressCloudSync();
}

/// Returns current XP (never null after [ensureXpInitialized]).
Future<int> getXp() async {
  final prefs = await SharedPreferences.getInstance();
  await ensureXpInitialized();
  return prefs.getInt(_prefUserXp) ?? defaultXp;
}

/// Adds [delta] XP (clamped so total is non-negative).
Future<int> addXp(int delta) async {
  if (delta <= 0) return await getXp();
  final prefs = await SharedPreferences.getInstance();
  final current = prefs.getInt(_prefUserXp) ?? defaultXp;
  final next = current + delta;
  await prefs.setInt(_prefUserXp, next);
  scheduleProgressCloudSync();
  return next;
}

/// Removes [delta] XP (clamped so total is non-negative).
Future<int> removeXp(int delta) async {
  if (delta <= 0) return await getXp();
  final prefs = await SharedPreferences.getInstance();
  final current = prefs.getInt(_prefUserXp) ?? defaultXp;
  final next = (current - delta).clamp(0, 0x7FFFFFFF);
  await prefs.setInt(_prefUserXp, next);
  scheduleProgressCloudSync();
  return next;
}

/// Consecutive frontier-level failures without clearing the level (persisted for skip-ad offer).
Future<int> getAdventureConsecutiveLosses() async {
  final prefs = await SharedPreferences.getInstance();
  return prefs.getInt(_prefAdventureConsecutiveLosses) ?? 0;
}

Future<void> resetAdventureConsecutiveLosses() async {
  final prefs = await SharedPreferences.getInstance();
  await prefs.setInt(_prefAdventureConsecutiveLosses, 0);
  scheduleProgressCloudSync();
}

/// Call when the player fails a frontier adventure level (loss popup). Returns the new streak.
Future<int> recordAdventureLevelFailure(int levelJustFailed) async {
  if (!await _isAdventureFrontierLevel(levelJustFailed)) {
    return getAdventureConsecutiveLosses();
  }
  final prefs = await SharedPreferences.getInstance();
  final next = (prefs.getInt(_prefAdventureConsecutiveLosses) ?? 0) + 1;
  await prefs.setInt(_prefAdventureConsecutiveLosses, next);
  scheduleProgressCloudSync();
  return next;
}

/// Adventure: call when player loses a game. Returns (new XP, delta). Applies -1 XP per loss.
Future<(int, int)> onAdventureMatchLost({int? levelJustPlayed}) async {
  await _ensureAdventureFirstClearXpMigrated();
  final prefs = await SharedPreferences.getInstance();
  await prefs.setInt(_prefAdventureConsecutiveWins, 0);

  final newXp = await removeXp(xpAdventureOneLoss);
  return (newXp, -xpAdventureOneLoss);
}

/// Adventure: call when player wins a game. Returns (new XP, delta).
/// +1 XP per win; +3 XP when clearing levels 10, 20, 30, …
Future<(int, int)> onAdventureGameWon({
  int? levelJustPlayed,
  bool levelClearingWin = false,
}) async {
  await _ensureAdventureFirstClearXpMigrated();
  final delta = levelClearingWin &&
          levelJustPlayed != null &&
          levelJustPlayed > 0 &&
          levelJustPlayed % 10 == 0
      ? xpAdventureMilestoneLevelWin
      : xpAdventureMatchWin;
  final newXp = await addXp(delta);
  return (newXp, delta);
}

/// Adventure: call when player wins the level (before advancing). Resets consecutive losses
/// and records first-clear tracking. Match wins already award XP via [onAdventureGameWon].
Future<(int, int)> onAdventureLevelWon(int levelJustCompleted) async {
  await _ensureAdventureFirstClearXpMigrated();
  final prefs = await SharedPreferences.getInstance();
  await prefs.setInt(_prefAdventureConsecutiveLosses, 0);

  final clearedList = prefs.getStringList(_prefAdventureLevelsFirstClearXp) ?? [];
  final levelKey = levelJustCompleted.toString();
  if (!clearedList.contains(levelKey)) {
    final nextList = [...clearedList, levelKey]
      ..sort((a, b) => int.parse(a).compareTo(int.parse(b)));
    await prefs.setStringList(_prefAdventureLevelsFirstClearXp, nextList);
  }
  scheduleProgressCloudSync();
  final xp = await getXp();
  return (xp, 0);
}

/// Classic: call when human wins in classic streak mode. Returns (new XP, delta). +2 if 3rd consecutive win.
Future<(int, int)> onClassicStreakWin(int classicGamesWonAfterThisWin) async {
  if (classicGamesWonAfterThisWin >= 3 && classicGamesWonAfterThisWin % 3 == 0) {
    final newXp = await addXp(xpClassicThreeWins);
    return (newXp, xpClassicThreeWins);
  }
  final xp = await getXp();
  return (xp, 0);
}

/// Practice: call when human wins a hard practice game. Returns (new XP, delta). +1 XP.
Future<(int, int)> onHardPracticeWin() async {
  final newXp = await addXp(xpHardPracticeWin);
  return (newXp, xpHardPracticeWin);
}

/// Rewarded ad on home / Gain XP — returns new XP total after [xpRewardedAdWatch].
Future<int> onRewardedAdWatched() async {
  return addXp(xpRewardedAdWatch);
}

/// Daily challenge: returns whether the user played today and if so whether they won.
/// (playedToday, wonOrNull). wonOrNull is null if not played today.
Future<(bool playedToday, bool? won)> getDailyChallengeStatus() async {
  final prefs = await SharedPreferences.getInstance();
  final now = DateTime.now();
  final today = '${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}';
  final lastDate = prefs.getString(_prefDailyChallengeDate);
  if (lastDate != today) return (false, null);
  final won = prefs.getBool(_prefDailyChallengeWon);
  return (true, won);
}

/// Daily challenge: call when the user finishes today's challenge. Records result only.
Future<int> setDailyChallengeResult(bool won) async {
  final prefs = await SharedPreferences.getInstance();
  final now = DateTime.now();
  final today = '${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')}';
  await prefs.setString(_prefDailyChallengeDate, today);
  await prefs.setBool(_prefDailyChallengeWon, won);
  scheduleProgressCloudSync();
  return await getXp();
}

/// Win or loss in a school lobby match (+1 / −1 XP).
Future<void> recordSchoolLobbyMatchOutcome(bool won) async {
  if (won) {
    await addXp(xpSchoolLobbyMatchDelta);
  } else {
    await removeXp(xpSchoolLobbyMatchDelta);
  }
}

/// Returns today's challenge game index (0-based). Same for all users on the same calendar day.
int getTodaysChallengeGameIndex() {
  final now = DateTime.now();
  final dayCode = now.year * 10000 + now.month * 100 + now.day;
  return dayCode % 6; // 6 different challenge types
}

/// Live Matches require at least 1 XP (send, accept, or rematch).
bool canPlayLiveMatches(int xp) => xp > 0;

const String liveMatchesRequiresXpMessage =
    'You need at least 1 XP to play Live Matches. Earn XP in Adventure or Classic mode, or watch an ad for +$xpRewardedAdWatch XP.';

File: bee-five-web/src/services/xpService.ts
import {
  DEFAULT_USER_XP,
  getProgressSyncUserId,
  readLocalPlayerStats,
  readLocalXpAuxState,
  scheduleProgressCloudSync,
  writeLocalPlayerStats,
  writeLocalXpAuxState,
} from './progressService';

const LAST_LOGIN_DATE_KEY = 'last_login_date';

export const defaultXp = DEFAULT_USER_XP;
export const xpClassicThreeWins = 2;
export const xpHardPracticeWin = 1;
export const xpAdventureOneLoss = 1;
export const xpAdventureTwoWins = 1;
export const xpAdventureFirstLevelComplete = 1;
export const xpAdventureMultipleOf10 = 5;
export const xpSchoolLobbyMatchDelta = 1;

export const liveMatchesRequiresXpMessage =
  'You need at least 1 XP to play Live Matches. Earn XP in Adventure or Classic mode.';

export type XpResult = { newXp: number; delta: number };

export interface AdventureXpContext {
  currentLevel: number;
  highestUnlocked: number;
}

const LOCAL_PROGRESS_KEY = 'beeAdventureProgress';

function canUseStorage(): boolean {
  return typeof window !== 'undefined' && typeof window.localStorage !== 'undefined';
}

function todayDateString(): string {
  const now = new Date();
  const month = String(now.getMonth() + 1).padStart(2, '0');
  const day = String(now.getDate()).padStart(2, '0');
  return `${now.getFullYear()}-${month}-${day}`;
}

function readAdventureProgressSnapshot(userId?: string | null): { current: number; highest: number } {
  if (!canUseStorage()) {
    return { current: 1, highest: 1 };
  }

  const keys = userId
    ? [`${LOCAL_PROGRESS_KEY}:${userId}`, `${LOCAL_PROGRESS_KEY}:guest`]
    : [`${LOCAL_PROGRESS_KEY}:guest`];

  for (const key of keys) {
    try {
      const raw = window.localStorage.getItem(key);
      if (!raw) continue;
      const progress = JSON.parse(raw) as {
        current_game?: number;
        highest_unlocked_game?: number;
      };
      const current = progress.current_game ?? 1;
      const highest = progress.highest_unlocked_game ?? current;
      return { current, highest: Math.max(current, highest) };
    } catch {
      // try next key
    }
  }

  return { current: 1, highest: 1 };
}

function effectiveAdventureFrontierLevel(ctx?: AdventureXpContext, userId?: string | null): number {
  if (ctx) {
    return Math.max(ctx.currentLevel, ctx.highestUnlocked);
  }
  const snap = readAdventureProgressSnapshot(userId);
  return Math.max(snap.current, snap.highest);
}

function isAdventureFrontierLevel(
  levelJustPlayedOrCompleted: number,
  ctx?: AdventureXpContext,
  userId?: string | null
): boolean {
  return levelJustPlayedOrCompleted === effectiveAdventureFrontierLevel(ctx, userId);
}

function currentUserId(): string | null {
  return getProgressSyncUserId();
}

function triggerCloudSync(): void {
  scheduleProgressCloudSync(currentUserId());
}

function ensureAdventureFirstClearXpMigrated(ctx?: AdventureXpContext, userId?: string | null): void {
  const resolvedUserId = userId ?? currentUserId();
  const aux = readLocalXpAuxState(resolvedUserId);
  if (aux.adventureFirstClearXpMigrated) return;

  const snap = ctx
    ? { current: ctx.currentLevel, highest: Math.max(ctx.currentLevel, ctx.highestUnlocked) }
    : readAdventureProgressSnapshot(resolvedUserId);
  const top = Math.max(snap.current, snap.highest);

  const existing = new Set(aux.adventureLevelsFirstClear.map(String));
  for (let i = 1; i < top; i++) {
    existing.add(String(i));
  }

  const list = Array.from(existing)
    .map((level) => Number.parseInt(level, 10))
    .filter((level) => Number.isFinite(level) && level > 0)
    .sort((a, b) => a - b);

  writeLocalXpAuxState(resolvedUserId, {
    adventureLevelsFirstClear: list,
    adventureFirstClearXpMigrated: true,
  });
  triggerCloudSync();
}

export function ensureXpInitialized(): void {
  if (!canUseStorage()) return;
  const userId = currentUserId();
  const stats = readLocalPlayerStats(userId);
  if (stats.userXp === DEFAULT_USER_XP) {
    writeLocalPlayerStats(userId, { userXp: DEFAULT_USER_XP });
  }
}

export function onAppOpen(): void {
  if (!canUseStorage()) return;
  ensureXpInitialized();

  const today = todayDateString();
  const last = window.localStorage.getItem(LAST_LOGIN_DATE_KEY);
  if (last === today) return;

  const userId = currentUserId();
  let streak = readLocalPlayerStats(userId).loginStreak;

  if (last == null) {
    streak = 1;
  } else {
    const lastDate = new Date(last);
    if (!Number.isNaN(lastDate.getTime())) {
      const now = new Date();
      const diffDays = Math.floor((now.getTime() - lastDate.getTime()) / (1000 * 60 * 60 * 24));
      if (diffDays === 1) {
        streak += 1;
      } else if (diffDays > 1) {
        streak = 1;
      }
    } else {
      streak = 1;
    }
  }

  writeLocalPlayerStats(userId, { loginStreak: streak });
  window.localStorage.setItem(LAST_LOGIN_DATE_KEY, today);
  triggerCloudSync();
}

export function getXp(): number {
  if (!canUseStorage()) return DEFAULT_USER_XP;
  ensureXpInitialized();
  return readLocalPlayerStats(currentUserId()).userXp;
}

export function addXp(delta: number): number {
  if (delta <= 0) return getXp();
  const userId = currentUserId();
  const next = getXp() + delta;
  writeLocalPlayerStats(userId, { userXp: next });
  triggerCloudSync();
  return next;
}

export function removeXp(delta: number): number {
  if (delta <= 0) return getXp();
  const userId = currentUserId();
  const next = Math.max(0, getXp() - delta);
  writeLocalPlayerStats(userId, { userXp: next });
  triggerCloudSync();
  return next;
}

export function canPlayLiveMatches(xp: number): boolean {
  return xp > 0;
}

export function onAdventureMatchLost(options?: {
  levelJustPlayed?: number;
  adventureContext?: AdventureXpContext;
  userId?: string | null;
}): XpResult {
  const resolvedUserId = options?.userId ?? currentUserId();
  ensureAdventureFirstClearXpMigrated(options?.adventureContext, resolvedUserId);
  writeLocalXpAuxState(resolvedUserId, { adventureConsecutiveWins: 0 });
  triggerCloudSync();

  const levelJustPlayed = options?.levelJustPlayed;
  if (levelJustPlayed != null) {
    const eligible = isAdventureFrontierLevel(
      levelJustPlayed,
      options?.adventureContext,
      resolvedUserId
    );
    if (!eligible) {
      return { newXp: getXp(), delta: 0 };
    }
  }

  const newXp = removeXp(xpAdventureOneLoss);
  return { newXp, delta: -xpAdventureOneLoss };
}

export function onAdventureGameWon(options?: {
  levelJustPlayed?: number;
  adventureContext?: AdventureXpContext;
  userId?: string | null;
}): XpResult {
  const resolvedUserId = options?.userId ?? currentUserId();
  ensureAdventureFirstClearXpMigrated(options?.adventureContext, resolvedUserId);

  const levelJustPlayed = options?.levelJustPlayed;
  if (levelJustPlayed != null) {
    const eligible = isAdventureFrontierLevel(
      levelJustPlayed,
      options?.adventureContext,
      resolvedUserId
    );
    if (!eligible) {
      return { newXp: getXp(), delta: 0 };
    }
  }

  const aux = readLocalXpAuxState(resolvedUserId);
  const wins = aux.adventureConsecutiveWins + 1;

  if (wins >= 2) {
    writeLocalXpAuxState(resolvedUserId, { adventureConsecutiveWins: 0 });
    const newXp = addXp(xpAdventureTwoWins);
    return { newXp, delta: xpAdventureTwoWins };
  }

  writeLocalXpAuxState(resolvedUserId, { adventureConsecutiveWins: wins });
  triggerCloudSync();
  return { newXp: getXp(), delta: 0 };
}

export function onAdventureLevelWon(
  levelJustCompleted: number,
  options?: {
    adventureContext?: AdventureXpContext;
    userId?: string | null;
  }
): XpResult {
  const resolvedUserId = options?.userId ?? currentUserId();
  ensureAdventureFirstClearXpMigrated(options?.adventureContext, resolvedUserId);

  const aux = readLocalXpAuxState(resolvedUserId);
  writeLocalXpAuxState(resolvedUserId, { adventureConsecutiveLosses: 0 });

  let delta = 0;
  const levelKey = levelJustCompleted;
  if (!aux.adventureLevelsFirstClear.includes(levelKey)) {
    const nextList = [...aux.adventureLevelsFirstClear, levelKey].sort((a, b) => a - b);
    writeLocalXpAuxState(resolvedUserId, { adventureLevelsFirstClear: nextList });
    addXp(xpAdventureFirstLevelComplete);
    delta += xpAdventureFirstLevelComplete;
  } else {
    triggerCloudSync();
  }

  const eligible = isAdventureFrontierLevel(
    levelJustCompleted,
    options?.adventureContext,
    resolvedUserId
  );
  if (!eligible) {
    return { newXp: getXp(), delta };
  }

  if (levelJustCompleted > 0 && levelJustCompleted % 10 === 0) {
    addXp(xpAdventureMultipleOf10);
    delta += xpAdventureMultipleOf10;
  }

  return { newXp: getXp(), delta };
}

export function onClassicStreakWin(classicGamesWonAfterThisWin: number): XpResult {
  if (classicGamesWonAfterThisWin >= 3 && classicGamesWonAfterThisWin % 3 === 0) {
    const newXp = addXp(xpClassicThreeWins);
    return { newXp, delta: xpClassicThreeWins };
  }
  return { newXp: getXp(), delta: 0 };
}

export function onHardPracticeWin(): XpResult {
  const newXp = addXp(xpHardPracticeWin);
  return { newXp, delta: xpHardPracticeWin };
}

export function getDailyChallengeStatus(): { playedToday: boolean; won: boolean | null } {
  const aux = readLocalXpAuxState(currentUserId());
  const today = todayDateString();
  if (aux.dailyChallengeDate !== today) return { playedToday: false, won: null };
  return { playedToday: true, won: aux.dailyChallengeWon };
}

export function setDailyChallengeResult(won: boolean): number {
  const userId = currentUserId();
  writeLocalXpAuxState(userId, {
    dailyChallengeDate: todayDateString(),
    dailyChallengeWon: won,
  });
  triggerCloudSync();
  return getXp();
}

export function recordSchoolLobbyMatchOutcome(won: boolean): void {
  if (won) addXp(xpSchoolLobbyMatchDelta);
  else removeXp(xpSchoolLobbyMatchDelta);
}

export function getTodaysChallengeGameIndex(): number {
  const now = new Date();
  const dayCode = now.getFullYear() * 10000 + (now.getMonth() + 1) * 100 + now.getDate();
  return dayCode % 6;
}
